src/pages/GradesPage.spec.tsx

```diff
@@ -0,0 +1,199 @@
+import {
+  render,
+  screen,
+  waitFor,
+  waitForElementToBeRemoved,
+} from "@testing-library/react";
+import { http, HttpResponse } from "msw";
+import { describe, expect, it, vi } from "vitest";
+import { emptyCourse, mockCourses, mockTerms } from "../tests/mocks/GradesMock";
+import { server } from "../tests/mocks/server";
+import { Providers } from "../tests/Providers";
+import GradesPage from "./GradesPage";
+import userEvent from "@testing-library/user-event";
+
+const setup = (guest = false) => {
+  const user = userEvent.setup();
+
+  render(
+    <Providers guest={guest}>
+      <GradesPage />
+    </Providers>
+  );
+
+  return { user };
+};
+
+describe("GradesPage", () => {
+  it("should show have restricted gui for guest users", async () => {
+    setup(true);
+
+    expect(screen.getByText(/oceny/i)).toBeInTheDocument();
+    expect(
+      screen.getByText(/nie jest dostępna w trybie gościa/i)
+    ).toBeInTheDocument();
+
+    const connectButton = screen.getByRole("button", { name: /połącz konto/i });
+    expect(connectButton).toBeInTheDocument();
+  });
+
+  it("should show loading spinner", async () => {
+    server.use(
+      http.get(
+        "/grades/",
+        () =>
+          new Promise((resolve) =>
+            setTimeout(() => resolve(HttpResponse.json({})), 200)
+          )
+      )
+    );
+    setup();
+
+    expect(screen.getByText(/ładowanie/i)).toBeInTheDocument();
+    await waitForElementToBeRemoved(() => screen.queryByText(/ładowanie/i));
+  });
+
+  it("should display fetched terms and courses", async () => {
+    vi.setSystemTime(new Date("2024-10-15T12:00:00Z"));
+    setup();
+
+    await waitFor(() => {
+      expect(
+        screen.getByRole("combobox", { name: /wybierz semestr/i })
+      ).toBeInTheDocument();
+
+      expect(screen.getByText(mockCourses[0].course_name)).toBeInTheDocument();
+      expect(screen.getByText(mockCourses[1].course_name)).toBeInTheDocument();
+      expect(
+        screen.getByRole("option", { name: mockTerms[0].name })
+      ).toBeInTheDocument();
+      expect(
+        screen.getByRole("option", { name: mockTerms[1].name })
+      ).toBeInTheDocument();
+    });
```

Comment:
trochę duży ten wait-for, raczej staraj się, żeby takie rzeczy były jak najmniejsze bo ukrywają co jest problemem